File: core/src/main/java/com/harera/hayat/service/user/UserService.java
package com.harera.hayat.service.user;

import org.modelmapper.ModelMapper;
import org.springframework.security.crypto.password.PasswordEncoder;

import com.harera.hayat.repository.UserRepository;
import com.harera.hayat.repository.user.auth.TokenRepository;
import com.harera.hayat.service.firebase.FirebaseService;
import com.harera.hayat.service.user.auth.AuthService;
import com.harera.hayat.service.user.auth.JwtService;
import com.harera.hayat.service.user.auth.JwtUtils;

public class UserService {

    private final UserRepository userRepository;
    private final AuthService authService;
    private final PasswordEncoder passwordEncoder;
    private final UserValidation userValidation;
    private final TokenRepository tokenRepository;
    private final FirebaseService firebaseService;
    private final ModelMapper modelMapper;
    private final JwtUtils jwtUtils;
    private final JwtService jwtService;
    private final UserUtils userUtils;

    public UserService(UserRepository userRepository, AuthService authService,
                    PasswordEncoder passwordEncoder, UserValidation userValidation,
                    TokenRepository tokenRepository, FirebaseService firebaseService,
                    ModelMapper modelMapper, JwtUtils jwtUtils, JwtService jwtService,
                    UserUtils userUtils) {
        this.userRepository = userRepository;
        this.authService = authService;
        this.passwordEncoder = passwordEncoder;
        this.userValidation = userValidation;
        this.tokenRepository = tokenRepository;
        this.firebaseService = firebaseService;
        this.modelMapper = modelMapper;
        this.jwtUtils = jwtUtils;
        this.jwtService = jwtService;
        this.userUtils = userUtils;
    }


}

File: src/main/java/com/harera/dwaaserver/service/user/UserService.kt
package com.harera.dwaaserver.service.user

import com.englizya.security.AuthenticationManager
import com.harera.dwaaserver.model.user.LoginRequest
import com.harera.dwaaserver.model.user.LoginResponse
import com.harera.dwaaserver.model.user.OAuthLoginRequest
import com.harera.dwaaserver.repository.UserRepository
import com.harera.dwaaserver.util.MStringUtils.isEmail
import com.harera.dwaaserver.util.MStringUtils.isPhoneNumber
import com.harera.dwaaserver.util.MStringUtils.isUsername
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service


interface UserService {

    fun login(loginRequest: LoginRequest): LoginResponse?
    fun login(oAuthLoginRequest: OAuthLoginRequest): LoginResponse?
}

@Service
open class UserServiceImpl(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
    private val userValidation: UserValidation,
    private val authenticationManager: AuthenticationManager,
) : UserService {

    override fun login(loginRequest: LoginRequest): LoginResponse? {
        userValidation.validateLogin(loginRequest)
        val uid = getUid(loginRequest.subject!!)
        val loginResponse =  LoginResponse(authenticationManager.generateToken(uid!!))
        return loginResponse
    }

    override fun login(oAuthLoginRequest: OAuthLoginRequest): LoginResponse? {
        userValidation.validateLogin(oAuthLoginRequest)
//        TODO("")
        return null
    }

    private fun getUid(subject: String): String? {
        return if (isPhoneNumber(subject)) {
            userRepository.findByPhoneNumber(subject).uid.toString()
        } else if (isEmail(subject)) {
            userRepository.findByEmail(subject).uid.toString()
        } else if (isUsername(subject)){
            userRepository.findByUsername(subject).uid.toString()
        } else {
            return null
        }
    }
}